File: VITRAL/vitral/src/vsdk/toolkit/media/RGBColorPalette.java
//===========================================================================
//=-------------------------------------------------------------------------=
//= Module history:                                                         =
//= - February 24 2006 - Gina Chiquillo: Original base version              =
//= - April 4 2006 - Oscar Chavarro: Quality check - comments added         =
//===========================================================================

package vsdk.toolkit.media;

import java.util.ArrayList;
import java.util.Iterator;
import vsdk.toolkit.common.ColorRgb;

public class RGBColorPalette {
    private ArrayList<ColorRgb> colors;

    public RGBColorPalette()
    {
        colors = new ArrayList<ColorRgb>();
        init(256);
    }

    public void init(int size)
    {
        colors = new ArrayList<ColorRgb>();
        int i;

        ColorRgb c;
        for ( i = 0; i < size; i++ ) {
            c = new ColorRgb(0, 0, 0);
            colors.add(c);
        }
        buildGrayLevelsTable();
    }

    public int size()
    {
        return colors.size();
    }

    public void buildGrayLevelsTable()
    {
        int pos = 0;
        double val = 0;

        for ( Iterator i = colors.iterator(); i.hasNext(); ) {
            ColorRgb c = (ColorRgb)i.next();
            c.r = c.g = c.b = val;
            val += 1.0/((double)(colors.size()-1));
            pos++;        
        }
    }

    public ColorRgb getColorAt(int i)
    {
        if ( i < 0 || i >= colors.size() ) return null;
        return (ColorRgb)colors.get(i);
    }

    public void setColorAt(int i, ColorRgb c)
    {
        if ( i < 0 || i >= colors.size() ) return;
        colors.set(i, c);
    }

    public void setColorAt(int i, double r, double g, double b)
    {
        if ( i < 0 || i >= colors.size() ) return;

        colors.set(i, new ColorRgb(r, g, b));
    }

    public void addColor(ColorRgb c)
    {
        colors.add(c);
    }

    public void addColor(double r, double g, double b)
    {
        colors.add(new ColorRgb(r, g, b));
    }

    /**
    The parameter `t` must be between 0 and 1
    */
    public ColorRgb evalNearest(double t)
    {
        if ( t < 0.0 ) t = 0.0;
        if ( t > 1.0 ) t = 1.0;

        int N = colors.size();

        int i = (int)(t*((double)N));

        if ( i < 0 ) i = 0;
        if ( i >= N ) i = N-1;

        return (ColorRgb)colors.get(i);
    }

    /**
    The parameter `t` must be between 0 and 1.
    */
    public ColorRgb evalLinear(double t)
    {
        if ( t < 0.0 ) t = 0.0;
        if ( t > 1.0 ) t = 1.0;

        int N = colors.size()-1;

        int inf = (int)(t*((double)N));
        int sup = inf+1;

        double delta = 1/((double)N);
        double r = t - inf*delta;
        double p = r / delta;

        if ( inf < 0 ) inf = 0;
        if ( inf > N ) inf = N;
        if ( sup < 0 ) sup = 0;
        if ( sup > N ) sup = N;

        ColorRgb a = colors.get(inf);
        ColorRgb b = colors.get(sup);
        ColorRgb c = new ColorRgb(
            a.r + (b.r-a.r)*p,
            a.g + (b.g-a.g)*p,
            a.b + (b.b-a.b)*p    );

        return c;
    }

}

//===========================================================================
//= EOF                                                                     =
//===========================================================================

File: cpp/base/src/main/vsdk/toolkit/media/RGBColorPalette.cpp
#include "vsdk/toolkit/media/RGBColorPalette.h"
#include "vsdk/toolkit/common/color/ColorRgb.h"
#include "java/util/ArrayList.txx"
#include <cmath>

RGBColorPalette::RGBColorPalette() {
    init(256);
}

RGBColorPalette::~RGBColorPalette() {
    for (long int i = 0; i < colors.size(); i++) {
        if (colors[i] != nullptr) {
            delete colors[i];
            colors[i] = nullptr;
        }
    }
    colors.clear();
}

void RGBColorPalette::init(int size) {
    for (long int i = 0; i < colors.size(); i++) {
        if (colors[i] != nullptr) {
            delete colors[i];
            colors[i] = nullptr;
        }
    }
    colors.clear();

    for (int i = 0; i < size; i++) {
        colors.add(new ColorRgb(0, 0, 0));
    }
    buildGrayLevelsTable();
}

int RGBColorPalette::size() const {
    return (int)colors.size();
}

void RGBColorPalette::buildGrayLevelsTable() {
    double val = 0.0;
    for (long int i = 0; i < colors.size(); i++) {
        if (colors[i] != nullptr) {
            delete colors[i];
        }
        colors[i] = new ColorRgb(val, val, val);
        val += 1.0 / ((double)(colors.size() - 1));
    }
}

ColorRgb* RGBColorPalette::getColorAt(int i) const {
    if (i < 0 || i >= (int)colors.size()) {
        return nullptr;
    }
    if (colors.get(i) != nullptr) {
        return new ColorRgb(*colors.get(i));
    }
    return nullptr;
}

void RGBColorPalette::setColorAt(int i, ColorRgb* c) {
    if (i < 0 || i >= (int)colors.size()) {
        return;
    }
    if (colors[i] != nullptr) {
        delete colors[i];
    }
    if (c != nullptr) {
        colors[i] = new ColorRgb(*c);
    } else {
        colors[i] = nullptr;
    }
}

void RGBColorPalette::setColorAt(int i, double r, double g, double b) {
    if (i < 0 || i >= (int)colors.size()) {
        return;
    }
    if (colors[i] != nullptr) {
        delete colors[i];
    }
    colors[i] = new ColorRgb(r, g, b);
}

void RGBColorPalette::addColor(ColorRgb* c) {
    if (c != nullptr) {
        colors.add(new ColorRgb(*c));
    } else {
        colors.add(new ColorRgb(0, 0, 0));
    }
}

void RGBColorPalette::addColor(double r, double g, double b) {
    colors.add(new ColorRgb(r, g, b));
}

ColorRgb* RGBColorPalette::evalNearest(double t) const {
    if (t < 0.0) t = 0.0;
    if (t > 1.0) t = 1.0;

    int N = (int)colors.size();
    int i = (int)(t * ((double)N));

    if (i < 0) i = 0;
    if (i >= N) i = N - 1;

    if (colors.get(i) != nullptr) {
        return new ColorRgb(*colors.get(i));
    }
    return new ColorRgb(0, 0, 0);
}

ColorRgb* RGBColorPalette::evalLinear(double t) const {
    if (t < 0.0) t = 0.0;
    if (t > 1.0) t = 1.0;

    int N = (int)colors.size() - 1;
    int inf = (int)(t * ((double)N));
    int sup = inf + 1;

    double delta = 1.0 / ((double)N);
    double r = t - inf * delta;
    double p = r / delta;

    if (inf < 0) inf = 0;
    if (inf > N) inf = N;
    if (sup < 0) sup = 0;
    if (sup > N) sup = N;

    ColorRgb* a = colors.get(inf);
    ColorRgb* b = colors.get(sup);

    double cr = a->r() + (b->r() - a->r()) * p;
    double cg = a->g() + (b->g() - a->g()) * p;
    double cb = a->b() + (b->b() - a->b()) * p;

    return new ColorRgb(cr, cg, cb);
}

int RGBColorPalette::selectNearestIndexToRgb(const ColorRgb& c) const {
    double minDistance = 1e308;
    int index = 0;

    for (long int i = 0; i < colors.size(); i++) {
        if (colors.get(i) != nullptr) {
            double currentDistance = colors.get(i)->distance(c);
            if (currentDistance < minDistance) {
                minDistance = currentDistance;
                index = (int)i;
            }
        }
    }
    return index;
}
